README
				GeoIP 1.5.1
				-----------

*** Experimental IPv6 notice ***

- the experimental IPv6 part of libGeoIP may change in the future.
- It is possible, that geoiplookup and geoiplookup6 will merged.

***

The new perl script geoipupdate-pureperl.pl is a drop in replacement for
geoipupdate. Useful, if you like to customize, pre or postprocess new or
old databases. To archive the databases or signal apache whatever.
Another advantage is, that geoipupdate-pureperl.pl is able to handle proxy
requests even with authentication.

IMPORTANT API Change for 1.3.x and above users for GeoIP Region database
GeoIPRegion.region is no longer a pointer but an in-structure
array so test the first byte of region == 0 rather testing if the region
pointer is NULL.

IMPORTANT API Change for 1.1.x and above users - as of GeoIP 1.1.0 the
GeoIP_country_xxx_by_xxx functions return NULL if a country can not
be found (it used to return '--' or 'N/A'. Be sure to check the
return value for NULL, to avoid segmentation faults!

GeoIP is a C library that enables the user to find geographical and
network information of an IP address.
Included is a free GeoLite Country database
that is updated at the beginning of every month.
To download the latest free GeoLite Country database, go to:
http://dev.maxmind.com/geoip/geolite

There is also a free city-level geolocation database, GeoLite City,
available from:
http://dev.maxmind.com/geoip/geolite

We also offer commercial GeoIP databases with greater accuracy and
additional network information, for more details, see:
https://www.maxmind.com/en/geolocation_landing

As of version 1.4.5 geoipupdate can handle updates via  HTTP Proxy Server.
If the environ variable http_proxy="http://proxy-host:port" is set.
The username:password (as in FTP URLs) is not supported! 
Thanks to Andrew Droffner for the patch!

As of version 1.3.6, the GeoIP C library is thread safe, as long as
GEOIP_CHECK_CACHE is not used.

If you use GeoIP to block access from high risk countries in order
to reduce fraud or abuse, you may want to also block access from known
proxy servers. For more details on our proxy detection service, please visit:
https://www.maxmind.com/en/proxy

To install, run:

./configure
make
make check
make install

The GeoIP C library relies on GNU make, not on BSD make

MEMORY CACHING AND OTHER OPTIONS

There are four options available:

GEOIP_STANDARD - read database from filesystem, uses least memory.

GEOIP_MEMORY_CACHE - load database into memory, faster performance
	but uses more memory

GEOIP_CHECK_CACHE - check for updated database. If database has been updated,
	reload filehandle and/or memory cache.

GEOIP_INDEX_CACHE - just cache
        the most frequently accessed index portion of the database, resulting
        in faster lookups than GEOIP_STANDARD, but less memory usage than
        GEOIP_MEMORY_CACHE - useful for larger databases such as
	GeoIP Organization and GeoIP City. Note, for GeoIP Country, Region
	and Netspeed databases, GEOIP_INDEX_CACHE is equivalent to GEOIP_MEMORY_CACHE

GEOIP_MMAP_CACHE - load database into mmap shared memory ( MMAP is not avail for WIN32 )

The options can be combined using bit operators. For example you can
use both GEOIP_MEMORY_CACHE and GEOIP_CHECK_CACHE by calling:

    GeoIP_open("/path/to/GeoIP.dat", GEOIP_MEMORY_CACHE | GEOIP_CHECK_CACHE);

By default, the city name is returned in iso-8859-1 charset. To obtain the
city name in utf8 instead, run:

    GeoIP_set_charset(gi, GEOIP_CHARSET_UTF8);

To get the netmask of the netblock of the last lookup, use GeoIP_last_netblock(gi).

EXAMPLES

See
test/
     test-geoip.c
     test-geoip-region.c
     test-geoip-city.c
     test-geoip-isp.c
     test-geoip-org.c
     test-geoip-netspeed.c

for examples of how to use the API. The test-geoip.c program works with both the GeoLite and
GeoIP Country databases. The test-geoip-city.c program works with both the GeoLite and
GeoIP City databases. The other example programs require the paid databases available
https://www.maxmind.com/en/geolocation_landing


AUTOMATIC UPDATES

MaxMind offers a service where you can have your database updated
automically each week. For more details see:

http://www.maxmind.com/en/license_key

RESOURCES Mailinglists

Please join the very low traffic mailinglists you are interested in.

http://sourceforge.net/mail/?group_id=66844

Preformance Patches.

Patrick McManus provide a patch to enhance the lookupspeed in MEMORY_CACHE mode. If you feel, that the current MEMORY_CACHE mode is to slow try the patch:

http://sourceforge.net/mailarchive/forum.php?forum_name=geoip-c-discuss&max_rows=25&style=nested&viewmonth=200803

RESOURCES Development version

Please find the latest version of the CAPI on github.

https://github.com/maxmind/geoip-api-c

TROUBLESHOOTING

In case of trouble building from source with libtool or autotools, run
./bootstrap or autoreconf -vfi or aclocal && autoconf && automake --add-missing
to update the generated configuration files.

The Windows build is not thread-safe in STANDARD mode b/c the pread is not
thread-safe.

If you run into trouble building your application with GeoIP support, try adding -fms-extensions to your CFLAGS. If you use Solaris and there C-Compiler use -features=extensions instead. These options enable unnamed union support and fix problems like: 'improper member use: dma_code' or 'GeoIPRecord' has no member named 'dma_code'.

Note that it is recommended that you use GNU make. Also, if you are using
OpenBSD, GeoIP requires OpenBSD 3.1 or greater.

if you get "cannot load shared object file: No such file or directory"
error, add the directory libGeoIP.so was installed to to /etc/ld.so.conf
and run ldconfig

On Solaris, if you get a
ld: fatal: relocations remain against allocatable but non-writable sections
error, try running

# make clean
# ./configure --disable-shared
# make

If you get a "ar : command not found" error, make sure that ar is
in your path. On Solaris, ar is typically found in /usr/ccs/bin

If you get a "geoipupdate.c:24: getopt.h: No such file or directory"
error, run

# export CPPFLAGS="-I/usr/local/include"

(assuming that getopt.h is in /usr/local/include)

If you get a "zlib.h: No such file or directory" error, make sure
that the zlib development libraries are installed on your server.
These are typically included in a "zlib-devel" package.

If you get a "bad interpreter: No such file or directory" error
when running ./configure, make sure that there are no DOS
returns in the configure script. To remove DOS returns,
run perl -pi -e 's!\r!!g' configure.

If gcc fails while consuming a large amount of memory, try
compiling with CFLAGS=-O1 (or -O0) instead of the default -O2.
It seems that some
versions of gcc have a bug and consume 1 GB of memory when optimizing
certain source files (the other source file where this was reported is
from XORG X-Server). It happens at least with gcc 3.3.1 and with gcc
4.2(.0). Thanks to Kai Schaetzl for the report.

If GEOIP_MMAP_CACHE doesn't work on a 64bit machine, try adding
the flag "MAP_32BIT" to the mmap call.

If you get a "passing argument 3 of 'gethostbyname_r' from incompatible pointer type"
error on AIX, download and/or untar a fresh copy of GeoIP. ( To avoid cached
results from a previous ./configure run )

cd ./GeoIP-1.4.6
then edit the file ./configure

and delete these two lines:

#define HAVE_GETHOSTBYNAME_R 1

#define GETHOSTBYNAME_R_RETURNS_INT 1

then save the configure script

and build it as usual

./configure
make
sudo make install



To submit a patch, please contact support@maxmind.com 


ChangeLog
1.7.0

* `GeoIP_string_delete` was added for freeing strings returned by the various
  lookup functions. This is useful on Windows when the library may have been
  compiled with one runtime and you are using another. Pull request by Simone
  Basso. GitHub #100.
* Fix build problems related to `rpl_malloc()`. Pull request by elig0n.
  GitHub #116.

1.6.12 2018-01-17

* Populate metro and area code when performing lookups in IPv6 City databases.
  Previously this was only done when using IPv4 City databases.

1.6.11 2017-05-15

* Fix use of a NULL pointer when opening a corrupt database with
  `GeoIP_open`. Reported by Stephan Zeisberg. GitHub #87.

1.6.10 2017-03-29

* GeoIP_database_info now returns the full version string rather than
  incorrectly truncating it. GitHub #79.
* This API is now distributed with a small test copy of GeoIP.dat rather than
  a full copy.
* Fix issue where Visual Studio 2015 was optimizing out initialization code.
  Reported and fixed by Scott Godin. GitHub #81.
* Fix test/benchmark on Windows. Gisle Vanem. GitHub #75.

1.6.9 2016-01-11

* Fix a regression introduced in version 1.6.8, which caused
  GeoIP_database_info to erroneously return NULL.

1.6.8 2016-01-11

* Allow compilation on older systems by relaxing the autoconf and automake
  minimum versions. Thank you, Jose Rubio!
* Avoid potential problems in multi-threaded environments by consistently using
  pread() rather than read().
* Fix various small issues reported by clang's static analyser.

1.6.7 2015-10-30

* Fixed a MSVC parser stack overflow when parsing `regionName.c` and
 `timeZone.c`. Fix by elliotlo. GitHub #64.
* Updated region codes and timezones.
* When using `GEOIP_MEMORY_CACHE` with an invalid database file, the search
  tree traversal could attempt to read memory outside of the memory allocated
  for the memory cache, resulting in a segmentation fault. A check was added
  to ensure that the traversal code does not try to read beyond the end of the
  file, whether in memory, memory mapped, or on disk.
* Previously the return values from file reads were ignored. We now check
  these values to ensure that there were no errors.


1.6.6 2015-07-28

* Replaced usage of deprecated fileno, read, and lseek on Visual Studio 2005+
  with their ISO C++ conformant replacements. (Fix by ClaudiuHKS. GitHub #55.)
* A warning about using a double as a float was fixed. (Fix by ClaudiuHKS.
  GitHub #56.)
* Fixed segfault when doing a lookup on an empty database. (Fix by NesoK.
  GitHub #62.)
* Fixed a memcheck error from valgrind in the `_check_mtime`
  function. (Reported by yurivct. GitHub #60.)
* Fixed `_check_mtime` to check the return value of `gettimeofday` rather than
  just assuming it worked.


1.6.5 2015-02-25

* A segmentation fault in `geoiplookup` was fixed when the utility was passed
  an invalid database. (Reported in Red Hat bug #1180874.)
* Additional validation was added for validation of the size used in the
  creation of the index cache. (Based on discussion in Red Hat bug #832913.)
* Changed the code to only look up country codes by using functions which
  ensure that we do not try to look past the end of an array. (Reported by
  Ivan Sorokin. GitHub #53)


1.6.4 2015-01-12

* Update Fips codes (Boris Zentner)
* Several issues with the MinGW build were fixed. (Thomas Pöchtrager. Github
  #46.)
* Use a constructor in pread.c to ensure the critical section is always
  initialized. (Thomas Pöchtrager. Github #47.)
* Added missing include of `io.h` on Windows. (Thomas Pöchtrager. Github #49.)
* Fixed `configure` warning that `'missing' script is too old or missing`.
  (Reported by Floren Munteanu. Github #33.)
* Previously `nmake /f Makefile.vc clean` would fail on Windows. This was
  fixed.
* Obsolete win32 and NetWare make files were removed.
* Numerous documentation updates. (Reported by Thomas Pöchtrager. GitHub #48.)


1.6.3 2014-10-29

* Added a GEOIP_SILENCE flag. Include this flag when calling GeoIP_open to
  prevent any messages from being written to stderr. (Philip Prindeville
  and Boris Zentner)
* Mitigate a possible race condition when running under threads in the

libGeoIP/GeoIP.h
/* -*- Mode: C; indent-tabs-mode: t; c-basic-offset: 2; tab-width: 2 -*- */
/* GeoIP.h
 *
 * Copyright (C) 2016 MaxMind, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
 */

#ifndef GEOIP_H
#define GEOIP_H

#ifdef __cplusplus
extern "C" {
#endif

#include <sys/types.h>
#if !defined(_WIN32)
#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/socket.h>
#else /* !defined(_WIN32) */
#include <winsock2.h>
#include <ws2tcpip.h>
#define FILETIME_TO_USEC(ft)                                                   \
    (((unsigned __int64)ft.dwHighDateTime << 32 | ft.dwLowDateTime) / 10)
#endif /* !defined(_WIN32) */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>  /* for fstat */
#include <sys/types.h> /* for fstat */

#define SEGMENT_RECORD_LENGTH 3
#define LARGE_SEGMENT_RECORD_LENGTH 4
#define STANDARD_RECORD_LENGTH 3
#define ORG_RECORD_LENGTH 4
#define MAX_RECORD_LENGTH 4
#define NUM_DB_TYPES (38 + 1)

/* 128 bit address in network order */
typedef struct in6_addr geoipv6_t;

#define GEOIP_CHKBIT_V6(bit, ptr)                                              \
    (ptr[((127UL - bit) >> 3)] & (1UL << (~(127 - bit) & 7)))

typedef struct GeoIPTag {
    FILE *GeoIPDatabase;
    char *file_path;
    unsigned char *cache;
    unsigned char *index_cache;
    unsigned int *databaseSegments;
    char databaseType;
    time_t mtime;
    int flags;
    off_t size;
    char record_length;
    int charset;     /* 0 iso-8859-1 1 utf8 */
    int record_iter; /* used in GeoIP_next_record */
    int netmask;     /* netmask of last lookup - set using depth in
                        _GeoIP_seek_record */
    time_t last_mtime_check;
    off_t dyn_seg_size; /* currently only used by the cityconfidence database */
    unsigned int ext_flags; /* bit 0 teredo support enabled */
} GeoIP;

typedef struct GeoIPLookup { int netmask; } GeoIPLookup;

typedef enum { GEOIP_TEREDO_BIT = 0 } GeoIPExtFlags;

typedef enum {
    GEOIP_CHARSET_ISO_8859_1 = 0,
    GEOIP_CHARSET_UTF8 = 1
} GeoIPCharset;

typedef struct GeoIPRegionTag {
    char country_code[3];
    char region[3];
} GeoIPRegion;

typedef enum {
    GEOIP_STANDARD = 0,
    GEOIP_MEMORY_CACHE = 1,
    GEOIP_CHECK_CACHE = 2,
    GEOIP_INDEX_CACHE = 4,
    GEOIP_MMAP_CACHE = 8,
    GEOIP_SILENCE = 16,
} GeoIPOptions;

typedef enum {
    GEOIP_COUNTRY_EDITION = 1,
    GEOIP_REGION_EDITION_REV0 = 7,
    GEOIP_CITY_EDITION_REV0 = 6,
    GEOIP_ORG_EDITION = 5,
    GEOIP_ISP_EDITION = 4,
    GEOIP_CITY_EDITION_REV1 = 2,
    GEOIP_REGION_EDITION_REV1 = 3,
    GEOIP_PROXY_EDITION = 8,
    GEOIP_ASNUM_EDITION = 9,
    GEOIP_NETSPEED_EDITION = 10,
    GEOIP_DOMAIN_EDITION = 11,
    GEOIP_COUNTRY_EDITION_V6 = 12,
    GEOIP_LOCATIONA_EDITION = 13,
    GEOIP_ACCURACYRADIUS_EDITION = 14,
    GEOIP_CITYCONFIDENCE_EDITION = 15,     /* unsupported */
    GEOIP_CITYCONFIDENCEDIST_EDITION = 16, /* unsupported */
    GEOIP_LARGE_COUNTRY_EDITION = 17,
    GEOIP_LARGE_COUNTRY_EDITION_V6 = 18,
    GEOIP_CITYCONFIDENCEDIST_ISP_ORG_EDITION =
        19,                         /* unused, but gaps are not allowed */
    GEOIP_CCM_COUNTRY_EDITION = 20, /* unused, but gaps are not allowed */
    GEOIP_ASNUM_EDITION_V6 = 21,
    GEOIP_ISP_EDITION_V6 = 22,
    GEOIP_ORG_EDITION_V6 = 23,
    GEOIP_DOMAIN_EDITION_V6 = 24,
    GEOIP_LOCATIONA_EDITION_V6 = 25,
    GEOIP_REGISTRAR_EDITION = 26,
    GEOIP_REGISTRAR_EDITION_V6 = 27,
    GEOIP_USERTYPE_EDITION = 28,
    GEOIP_USERTYPE_EDITION_V6 = 29,
    GEOIP_CITY_EDITION_REV1_V6 = 30,
    GEOIP_CITY_EDITION_REV0_V6 = 31,
    GEOIP_NETSPEED_EDITION_REV1 = 32,
    GEOIP_NETSPEED_EDITION_REV1_V6 = 33,
    GEOIP_COUNTRYCONF_EDITION = 34,
    GEOIP_CITYCONF_EDITION = 35,
    GEOIP_REGIONCONF_EDITION = 36,
    GEOIP_POSTALCONF_EDITION = 37,
    GEOIP_ACCURACYRADIUS_EDITION_V6 = 38
} GeoIPDBTypes;

typedef enum {
    GEOIP_ANON_PROXY = 1,
    GEOIP_HTTP_X_FORWARDED_FOR_PROXY = 2,
    GEOIP_HTTP_CLIENT_IP_PROXY = 3,
} GeoIPProxyTypes;

typedef enum {
    GEOIP_UNKNOWN_SPEED = 0,
    GEOIP_DIALUP_SPEED = 1,
    GEOIP_CABLEDSL_SPEED = 2,
    GEOIP_CORPORATE_SPEED = 3,
} GeoIPNetspeedValues;

#if defined(_WIN32) && !defined(__MINGW32__)
#ifdef GEOIP_EXPORTS
#define GEOIP_API __declspec(dllexport)
#define GEOIP_DATA __declspec(dllexport)
#elif GEOIP_STATIC
#define GEOIP_API
#define GEOIP_DATA
#else
#define GEOIP_DATA __declspec(dllimport)
#define GEOIP_API
#endif /* GEOIP_EXPORTS */
#else
#define GEOIP_API
#define GEOIP_DATA
#endif

extern char **GeoIPDBFileName;
extern GEOIP_DATA const char *GeoIPDBDescription[NUM_DB_TYPES];
extern GEOIP_DATA const char *GeoIPCountryDBFileName;
extern GEOIP_DATA const char *GeoIPRegionDBFileName;
extern GEOIP_DATA const char *GeoIPCityDBFileName;
extern GEOIP_DATA const char *GeoIPOrgDBFileName;
extern GEOIP_DATA const char *GeoIPISPDBFileName;
extern GEOIP_DATA const char *GeoIPLocationADBFileName;
extern GEOIP_DATA const char *GeoIPAccuracyRadiusFileName;
extern GEOIP_DATA const char *GeoIPCityConfidenceFileName;
extern char *GeoIP_custom_directory;

/* Warning: do not use those arrays as doing so may break your
 * program with newer GeoIP versions */
extern GEOIP_DATA const char GeoIP_country_code[256][3];
extern GEOIP_DATA const char GeoIP_country_code3[256][4];
extern GEOIP_DATA const char *GeoIP_country_name[256];
extern GEOIP_DATA const char *GeoIP_utf8_country_name[256];
extern GEOIP_DATA const char GeoIP_country_continent[256][3];

GEOIP_API void GeoIP_setup_custom_directory(char *dir);
GEOIP_API GeoIP *GeoIP_open_type(int type, int flags);
GEOIP_API GeoIP *GeoIP_new(int flags);
GEOIP_API GeoIP *GeoIP_open(const char *filename, int flags);
/*
 * WARNING: GeoIP_db_avail() checks for the existence of a database
 * file but does not check that it has the requested database revision.
 * For database types which have more than one revision (including Region,
 * City, and Cityv6), this can lead to unexpected results. Check the
 * return value of GeoIP_open_type() to find out whether a particular
 * database type is really available.
 */
GEOIP_API int GeoIP_db_avail(int type);
GEOIP_API void GeoIP_delete(GeoIP *gi);

GEOIP_API const char *
GeoIP_country_code_by_addr_gl(GeoIP *gi, const char *addr, GeoIPLookup *gl);
GEOIP_API const char *
GeoIP_country_code_by_name_gl(GeoIP *gi, const char *host, GeoIPLookup *gl);
GEOIP_API const char *
GeoIP_country_code3_by_addr_gl(GeoIP *gi, const char *addr, GeoIPLookup *gl);
GEOIP_API const char *
GeoIP_country_code3_by_name_gl(GeoIP *gi, const char *host, GeoIPLookup *gl);
GEOIP_API const char *
GeoIP_country_name_by_addr_gl(GeoIP *gi, const char *addr, GeoIPLookup *gl);
GEOIP_API const char *
GeoIP_country_name_by_name_gl(GeoIP *gi, const char *host, GeoIPLookup *gl);
GEOIP_API const char *
GeoIP_country_name_by_ipnum_gl(GeoIP *gi, unsigned long ipnum, GeoIPLookup *gl);
GEOIP_API const char *
GeoIP_country_code_by_ipnum_gl(GeoIP *gi, unsigned long ipnum, GeoIPLookup *gl);
GEOIP_API const char *GeoIP_country_code3_by_ipnum_gl(GeoIP *gi,
                                                      unsigned long ipnum,
                                                      GeoIPLookup *gl);

/* */
GEOIP_API const char *
GeoIP_country_name_by_ipnum_v6_gl(GeoIP *gi, geoipv6_t ipnum, GeoIPLookup *gl);
GEOIP_API const char *
GeoIP_country_code_by_ipnum_v6_gl(GeoIP *gi, geoipv6_t ipnum, GeoIPLookup *gl);
GEOIP_API const char *
GeoIP_country_code3_by_ipnum_v6_gl(GeoIP *gi, geoipv6_t ipnum, GeoIPLookup *gl);

GEOIP_API const char *
GeoIP_country_code_by_addr_v6_gl(GeoIP *gi, const char *addr, GeoIPLookup *gl);
GEOIP_API const char *
GeoIP_country_code_by_name_v6_gl(GeoIP *gi, const char *host, GeoIPLookup *gl);
GEOIP_API const char *
GeoIP_country_code3_by_addr_v6_gl(GeoIP *gi, const char *addr, GeoIPLookup *gl);
GEOIP_API const char *
GeoIP_country_code3_by_name_v6_gl(GeoIP *gi, const char *host, GeoIPLookup *gl);
GEOIP_API const char *
GeoIP_country_name_by_addr_v6_gl(GeoIP *gi, const char *addr, GeoIPLookup *gl);
GEOIP_API const char *
GeoIP_country_name_by_name_v6_gl(GeoIP *gi, const char *host, GeoIPLookup *gl);

GEOIP_API int GeoIP_id_by_addr_gl(GeoIP *gi, const char *addr, GeoIPLookup *gl);
GEOIP_API int GeoIP_id_by_name_gl(GeoIP *gi, const char *host, GeoIPLookup *gl);
GEOIP_API int
GeoIP_id_by_ipnum_gl(GeoIP *gi, unsigned long ipnum, GeoIPLookup *gl);

GEOIP_API int
GeoIP_id_by_addr_v6_gl(GeoIP *gi, const char *addr, GeoIPLookup *gl);
GEOIP_API int
GeoIP_id_by_name_v6_gl(GeoIP *gi, const char *host, GeoIPLookup *gl);
GEOIP_API int
GeoIP_id_by_ipnum_v6_gl(GeoIP *gi, geoipv6_t ipnum, GeoIPLookup *gl);

GEOIP_API GeoIPRegion *
GeoIP_region_by_addr_gl(GeoIP *gi, const char *addr, GeoIPLookup *gl);
GEOIP_API GeoIPRegion *
GeoIP_region_by_name_gl(GeoIP *gi, const char *host, GeoIPLookup *gl);
GEOIP_API GeoIPRegion *
GeoIP_region_by_ipnum_gl(GeoIP *gi, unsigned long ipnum, GeoIPLookup *gl);

GEOIP_API GeoIPRegion *
GeoIP_region_by_addr_v6_gl(GeoIP *gi, const char *addr, GeoIPLookup *gl);
GEOIP_API GeoIPRegion *
GeoIP_region_by_name_v6_gl(GeoIP *gi, const char *host, GeoIPLookup *gl);
GEOIP_API GeoIPRegion *
GeoIP_region_by_ipnum_v6_gl(GeoIP *gi, geoipv6_t ipnum, GeoIPLookup *gl);

/* Warning - don't call this after GeoIP_assign_region_by_inetaddr calls */
GEOIP_API void GeoIPRegion_delete(GeoIPRegion *gir);

GEOIP_API void GeoIP_assign_region_by_inetaddr_gl(GeoIP *gi,
                                                  unsigned long inetaddr,
                                                  GeoIPRegion *gir,
                                                  GeoIPLookup *gl);
GEOIP_API void GeoIP_assign_region_by_inetaddr_v6_gl(GeoIP *gi,
                                                     geoipv6_t inetaddr,
                                                     GeoIPRegion *gir,
                                                     GeoIPLookup *gl);

/* Used to query GeoIP Organization, ISP and AS Number databases */
GEOIP_API char *
GeoIP_name_by_ipnum_gl(GeoIP *gi, unsigned long ipnum, GeoIPLookup *gl);
GEOIP_API char *
GeoIP_name_by_addr_gl(GeoIP *gi, const char *addr, GeoIPLookup *gl);
GEOIP_API char *
GeoIP_name_by_name_gl(GeoIP *gi, const char *host, GeoIPLookup *gl);

GEOIP_API char *
GeoIP_name_by_ipnum_v6_gl(GeoIP *gi, geoipv6_t ipnum, GeoIPLookup *gl);
GEOIP_API char *
GeoIP_name_by_addr_v6_gl(GeoIP *gi, const char *addr, GeoIPLookup *gl);
GEOIP_API char *
GeoIP_name_by_name_v6_gl(GeoIP *gi, const char *name, GeoIPLookup *gl);

/*
 * Free a malloc()-ed string returned by other APIs. Handy to have on
 * Windows where you may have different libraries linked with different
 * versions of the runtime (e.g. MT debug and MT release).
 */
GEOIP_API void GeoIP_string_delete(char *ptr);

/** return two letter country code */
GEOIP_API const char *GeoIP_code_by_id(int id);

/** return three letter country code */
GEOIP_API const char *GeoIP_code3_by_id(int id);

/** return full name of country in utf8 or iso-8859-1 */
GEOIP_API const char *GeoIP_country_name_by_id(GeoIP *gi, int id);

/** return full name of country */
GEOIP_API const char *GeoIP_name_by_id(int id);

/** return continent of country */
GEOIP_API const char *GeoIP_continent_by_id(int id);

/** return id by country code **/
GEOIP_API int GeoIP_id_by_code(const char *country);

/** return return number of known countries */
GEOIP_API unsigned GeoIP_num_countries(void);

GEOIP_API char *GeoIP_database_info(GeoIP *gi);
GEOIP_API unsigned char GeoIP_database_edition(GeoIP *gi);

GEOIP_API int GeoIP_charset(GeoIP *gi);
GEOIP_API int GeoIP_set_charset(GeoIP *gi, int charset);
GEOIP_API int GeoIP_enable_teredo(GeoIP *gi, int true_false);
GEOIP_API int GeoIP_teredo(GeoIP *gi);

GEOIP_API char **
GeoIP_range_by_ip_gl(GeoIP *gi, const char *addr, GeoIPLookup *gl);
GEOIP_API void GeoIP_range_by_ip_delete(char **ptr);

/* Convert region code to region name */
GEOIP_API const char *GeoIP_region_name_by_code(const char *country_code,
                                                const char *region_code);

/* Get timezone from country and region code */
GEOIP_API const char *
GeoIP_time_zone_by_country_and_region(const char *country_code,
                                      const char *region_code);

/* some v4 helper functions as of 1.4.7 exported to the public API */
GEOIP_API unsigned long GeoIP_addr_to_num(const char *addr);
GEOIP_API char *GeoIP_num_to_addr(unsigned long ipnum);

/* Internal function -- convert iso to utf8; return a malloced utf8 string. */
char *_GeoIP_iso_8859_1__utf8(const char *iso);

/* Cleans up memory used to hold file name paths. Returns 1 if successful;
 * otherwise 0.
 * */
GEOIP_API int GeoIP_cleanup(void);

/* Returns the library version in use. Helpful if you're loading dynamically. */
GEOIP_API const char *GeoIP_lib_version(void);

/* deprecated */
GEOIP_API const char *GeoIP_country_code_by_addr(GeoIP *gi, const char *addr);
GEOIP_API const char *GeoIP_country_code_by_name(GeoIP *gi, const char *host);
GEOIP_API const char *GeoIP_country_code3_by_addr(GeoIP *gi, const char *addr);
GEOIP_API const char *GeoIP_country_code3_by_name(GeoIP *gi, const char *host);
GEOIP_API const char *GeoIP_country_name_by_addr(GeoIP *gi, const char *addr);
GEOIP_API const char *GeoIP_country_name_by_name(GeoIP *gi, const char *host);
GEOIP_API const char *GeoIP_country_name_by_ipnum(GeoIP *gi,
                                                  unsigned long ipnum);
GEOIP_API const char *GeoIP_country_code_by_ipnum(GeoIP *gi,
                                                  unsigned long ipnum);
GEOIP_API const char *GeoIP_country_code3_by_ipnum(GeoIP *gi,
                                                   unsigned long ipnum);

GEOIP_API const char *GeoIP_country_name_by_ipnum_v6(GeoIP *gi,
                                                     geoipv6_t ipnum);
GEOIP_API const char *GeoIP_country_code_by_ipnum_v6(GeoIP *gi,
                                                     geoipv6_t ipnum);
GEOIP_API const char *GeoIP_country_code3_by_ipnum_v6(GeoIP *gi,
                                                      geoipv6_t ipnum);

GEOIP_API const char *GeoIP_country_code_by_addr_v6(GeoIP *gi,
                                                    const char *addr);
GEOIP_API const char *GeoIP_country_code_by_name_v6(GeoIP *gi,
                                                    const char *host);
GEOIP_API const char *GeoIP_country_code3_by_addr_v6(GeoIP *gi,
                                                     const char *addr);
GEOIP_API const char *GeoIP_country_code3_by_name_v6(GeoIP *gi,
                                                     const char *host);
GEOIP_API const char *GeoIP_country_name_by_addr_v6(GeoIP *gi,
                                                    const char *addr);
GEOIP_API const char *GeoIP_country_name_by_name_v6(GeoIP *gi,
                                                    const char *host);

GEOIP_API int GeoIP_id_by_addr(GeoIP *gi, const char *addr);
GEOIP_API int GeoIP_id_by_name(GeoIP *gi, const char *host);
GEOIP_API int GeoIP_id_by_ipnum(GeoIP *gi, unsigned long ipnum);

GEOIP_API int GeoIP_id_by_addr_v6(GeoIP *gi, const char *addr);
GEOIP_API int GeoIP_id_by_name_v6(GeoIP *gi, const char *host);
GEOIP_API int GeoIP_id_by_ipnum_v6(GeoIP *gi, geoipv6_t ipnum);

GEOIP_API GeoIPRegion *GeoIP_region_by_addr(GeoIP *gi, const char *addr);
GEOIP_API GeoIPRegion *GeoIP_region_by_name(GeoIP *gi, const char *host);
GEOIP_API GeoIPRegion *GeoIP_region_by_ipnum(GeoIP *gi, unsigned long ipnum);

GEOIP_API GeoIPRegion *GeoIP_region_by_addr_v6(GeoIP *gi, const char *addr);
GEOIP_API GeoIPRegion *GeoIP_region_by_name_v6(GeoIP *gi, const char *host);
GEOIP_API GeoIPRegion *GeoIP_region_by_ipnum_v6(GeoIP *gi, geoipv6_t ipnum);

GEOIP_API void GeoIP_assign_region_by_inetaddr(GeoIP *gi,
                                               unsigned long inetaddr,
                                               GeoIPRegion *gir);
GEOIP_API void GeoIP_assign_region_by_inetaddr_v6(GeoIP *gi,
                                                  geoipv6_t inetaddr,
                                                  GeoIPRegion *gir);

GEOIP_API char *GeoIP_name_by_ipnum(GeoIP *gi, unsigned long ipnum);
GEOIP_API char *GeoIP_name_by_addr(GeoIP *gi, const char *addr);
GEOIP_API char *GeoIP_name_by_name(GeoIP *gi, const char *host);

GEOIP_API char *GeoIP_name_by_ipnum_v6(GeoIP *gi, geoipv6_t ipnum);
GEOIP_API char *GeoIP_name_by_addr_v6(GeoIP *gi, const char *addr);
GEOIP_API char *GeoIP_name_by_name_v6(GeoIP *gi, const char *name);

/** GeoIP_last_netmask is deprecated - it is not thread safe */
GEOIP_API int GeoIP_last_netmask(GeoIP *gi);
GEOIP_API char **GeoIP_range_by_ip(GeoIP *gi, const char *addr);

/* Deprecated - for backwards compatibility only */
GEOIP_API int GeoIP_country_id_by_addr(GeoIP *gi, const char *addr);
GEOIP_API int GeoIP_country_id_by_name(GeoIP *gi, const char *host);
GEOIP_API char *GeoIP_org_by_addr(GeoIP *gi, const char *addr);
GEOIP_API char *GeoIP_org_by_name(GeoIP *gi, const char *host);
GEOIP_API char *GeoIP_org_by_ipnum(GeoIP *gi, unsigned long ipnum);

GEOIP_API int GeoIP_country_id_by_addr_v6(GeoIP *gi, const char *addr);
GEOIP_API char *GeoIP_org_by_ipnum_v6(GeoIP *gi, geoipv6_t ipnum);
GEOIP_API char *GeoIP_org_by_addr_v6(GeoIP *gi, const char *addr);
GEOIP_API char *GeoIP_org_by_name_v6(GeoIP *gi, const char *name);

/* End deprecated */

#
#ifdef __cplusplus
}
#endif

#endif /* GEOIP_H */
